marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/job/startJob.sjs

```diff
@@ -14,24 +14,39 @@
  limitations under the License.
  */
 'use strict';
-
 xdmp.securityAssert("http://marklogic.com/data-hub/privileges/run-step", "execute");
 
 const consts = require('/data-hub/5/impl/consts.sjs');
+const config = require("/com.marklogic.hub/config.sjs");
 const hubUtils = require("/data-hub/5/impl/hub-utils.sjs");
 const Job = require("/data-hub/5/flow/job.sjs");
 const jobs = require("/data-hub/5/impl/jobs.sjs");
+const dhProv = require('/data-hub/5/provenance/dh-provenance.xqy');
 
 var jobId;
 var flowName;
 
+function recordJobProvenance(jobId, timeStarted) {
+  const options = {
+    "startTime": timeStarted,
+    "user": xdmp.getCurrentUser()
+  }
+  const record = dhProv.provenanceRecord(jobId, options);
+  dhProv.insertProvenanceRecord(record, config.FINALDATABASE);
+  dhProv.insertProvenanceRecord(record, config.STAGINGDATABASE);
```

Comment:
This doesn't always seem appropriate - i.e. should this happen for a job that only writes to one database? The story doesn't explain what to do either.

Reply:
Added the information to the story. The job provenance record is inserted to both the staging and final content databases during the job start as we are not aware of the target database. The unused provenance record will be deleted when the job finishes. The cleanup will be addressed as part of another ticket.

Reply:
Hmm - only a step has a target database, not a flow. Since a job can involve running multiple steps, a job may involve writing data to multiple databases. It is conceivable that step 1 writes to staging, step 2 writes to final, and step 3 writes to staging. What would the target database be in that scenario?

Reply:
The job provenance record is initially written to both staging and final databases
1) step1 writes to staging, step2 writes to final -- then the records in both staging and final exists. If a step target database is staging, provenance is written into the record created in the staging database. If it is final database, then the provenance is written to the record created in the final database
2) step1 writes to final, step2 writes to final -- then the record in final exists and staging will be deleted and vice-versa
This is yet to be implemented. Will be implemented in the future when writing provenance for each step.

Reply:
Okay - though would it make life easier if the call to startJob included the first step to be run, such that the endpoint could figure out the target database of the first step and thus avoid writing a document that needs to be deleted later? Note that there's no guarantee that that document will get deleted, as the user could stop the flow at any time.

Reply:
1) It will be easy if startJob included all the steps that the user wants to run
2) When user stops a flow, will the finishJob be called?  I am assuming that the job document will be updated when the user stops running the flow by calling finishJob which will also clean up the provenance records

Reply:
If a user stops a flow run by Gradle or the client jar via ctrl-c, then the JVM will terminate immediately and finishJob won't be called. For that reason, I think it's better if we only create a job provenance document when we know we need one. 

As for passing along the first step, or any of them - the initializeFlow method in FlowRunnerImpl calls startJob. That code can get the steps from stepsMap.get(jobId). So you can then modify the startJob DS endpoint to accept any other data that you need, whether it's the first step to be run or all of the step numbers. The startJob DS endpoint can then look up the full flow definition to determine that the target databases are.

Reply:
For example, if I all the step numbers are used to create the provenance records, then the issue with stopping the flow using Gradle or the client jar via ctrl-c still exists.

So its better to create the prov record based on first step. When recording job prov information for a step, we can create the job prov record if necessary or update an existing record. Am I right?